src/lib.rs

```diff
@@ -1311,6 +1312,118 @@ where
     Some(Ratio::new(n1, d1))
 }
 
+impl<T: Clone + Integer + Signed + ToPrimitive + ToBigInt> Ratio<T> {
+    /// Converts the ratio to an `f64`.
+    ///
+    /// The value is guaranteed to be the closest `f64` approximating the ratio. This always
+    /// returns a value, unless the conversion to BigInt for the numerator or denominator fails.
+    ///
+    /// ```
+    /// use num_rational::Rational64;
+    /// assert_eq!(0.5f64, Rational64::new(1, 2).to_f64().unwrap());
+    /// ```
+    pub fn to_f64(&self) -> Option<f64> {
+        assert_eq!(
+            std::f64::RADIX,
+            2,
+            "only floating point implementations with radix 2 are currently supported"
+        );
+        assert!(
+            std::mem::size_of::<isize>() >= 4,
+            "platforms 16 bits and under are not supported"
+        );
+
+        // The following casts all work on 32-bit+ platforms, but may overflow on 16-bit ones.
+        const F64_MAX_EXP: isize = std::f64::MAX_EXP as isize;
+        const F64_MIN_EXP: isize = std::f64::MIN_EXP as isize;
+        const F64_MANTISSA_DIGITS: isize = std::f64::MANTISSA_DIGITS as isize;
+
+        // Upper bound to the range of exactly-representable ints in an f64.
+        const MAX_EXACT_INT: u64 = 1u64 << std::f64::MANTISSA_DIGITS;
+
+        let numer: T = self.numer().abs();
+        let denom: T = self.denom().abs();
+        let sign: T = self.numer().signum() * self.denom().signum();
+        let flo_sign = if sign.is_negative() { -1.0 } else { 1.0 };
+
+        if numer.is_zero() {
+            return Some(0.0 * flo_sign);
+        }
+
+        // Fast track: both sides can losslessly be converted to f64s. In this case, letting the
+        // FPU do the job is faster and easier. In any other case, converting to f64s may lead
+        // to an inexact result: https://stackoverflow.com/questions/56641441/
+        if numer
+            .to_u64()
+            .and_then(|n| {
+                denom
+                    .to_u64()
+                    .map(|d| n < MAX_EXACT_INT && d < MAX_EXACT_INT)
+            })
+            .unwrap_or(false)
+        {
+            return Some(
+                numer
+                    .to_f64()
+                    .and_then(|n| denom.to_f64().map(|d| n / d * flo_sign))
+                    .unwrap(),
+            );
+        }
+
+        // Otherwise, the goal is to obtain a quotient with at least 55 bits. 53 of these bits will
+        // be used as the mantissa of the resulting float, and the remaining two are for rounding.
+        // There's an error of up to 1 on the number of resulting bits, so we may get either 55 or
+        // 56 bits.
+        let mut numer = numer.to_bigint()?;
+        let denom = denom.to_bigint()?;
+        let diff = numer.bits() as isize - denom.bits() as isize;
```

Comment:
this fails when `numer.bits()` is greater than the maximum `isize`, but smaller than the maximum `usize`. You could refactor so diff is a tuple of `(bool, usize)` where `bool` indicates which is bigger and `usize` indicates by how much.